### PR TITLE
Create the basic structure of device doctor tool

### PR DESCRIPTION
**Pull request by keyonghan**
https://github.com/flutter/cocoon/pull/1016 is way too big to make it difficult to review (sorry for that 😢 ).

Here come the divided pieces, and this is the first one for basic structure. It addressed some comments from the above PR. Please help take another look.

What's included in this PR:
- main.dart to include only args check and initialization
- includes only ios device case for simplicity
- adds basic utils-support for current logic.
- creates fake ios device for testing
- adds corresponding tests

More PRs to come to finish up the whole tool.

Related flutter/flutter#66193

### CODE REVIEW

File: device_doctor/bin/main.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:args/args.dart';
+
+import 'package:device_doctor/src/config.dart';
```
Comment (CaseyHillers) [inline device_doctor/bin/main.dart:9]: Prefer relative imports - https://dart-lang.github.io/linter/lints/prefer_relative_imports.html

File: device_doctor/bin/main.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+const String actionFlag = 'action';
+const String deviceOSFlag = 'deviceOS';
```
Comment (CaseyHillers) [inline device_doctor/bin/main.dart:12]: nit: CLI args in the Flutter tooling follows the style of `--device-os`

File: device_doctor/bin/main.dart [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  if (!_checkArgs(parser, args)) {
+    stdout.write('\nRequired flags:\n'
+        '--$actionFlag Supported actions are healthcheck and recovery.\n'
+        '--$deviceOSFlag Supported device OS: android and ios.\n');
+    exit(1);
+  }
+
+  Config.initialize(_deviceOS);
+
+  // Healthcheck and recovery will be implemented here. PR to be continued.
```
Comment (CaseyHillers) [inline device_doctor/bin/main.dart:40]: Is there an issue we can track to?

File: device_doctor/bin/main.dart [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+Future<void> main(List<String> args) async {
+  final ArgParser parser = ArgParser();
+  parser
+    ..addOption('$actionFlag')
+    ..addOption('$deviceOSFlag')
+    ..addFlag('$helpFlag');
+
+  if (!_checkArgs(parser, args)) {
+    stdout.write('\nRequired flags:\n'
+        '--$actionFlag Supported actions are healthcheck and recovery.\n'
```
Comment (CaseyHillers) [inline device_doctor/bin/main.dart:33]: Arg parser can be used to generate this dynamically. The devicelab test runner does it that way - https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/run.dart

File: device_doctor/lib/src/config.dart [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import 'device.dart';
+
+Config _config;
+Config get config => _config;
+
+/// Configures device operating system based on injected decice OS.
```
Comment (CaseyHillers) [inline device_doctor/lib/src/config.dart:12]: nit: decice -> device

File: device_doctor/lib/src/config.dart [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+///
+/// This determines the [DeviceDiscovery].
+class Config {
```
Comment (CaseyHillers) [inline device_doctor/lib/src/config.dart:15]: This doesn't need to be addressed this PR, but we should consider passing a device ip instead of assuming there is only ever going to be one device attached.

File: device_doctor/lib/src/device.dart [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+enum DeviceOperatingSystem { android, ios }
+
+/// Discovers available devices and chooses one to work with.
+abstract class DeviceDiscovery {
+  factory DeviceDiscovery() {
+    switch (config.deviceOperatingSystem) {
+      case DeviceOperatingSystem.ios:
+        return IosDeviceDiscovery();
+      default:
+        throw StateError('Unsupported device operating system: {config.deviceOperatingSystem}');
```
Comment (CaseyHillers) [inline device_doctor/lib/src/device.dart:23]: nit: is this missing $? ${config.deviceOperatingSystem}

File: device_doctor/lib/src/device.dart [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+
+  /// Lists all available devices' IDs.
+  Future<List<Device>> discoverDevices({int retriesDelayMs = 10000});
```
Comment (CaseyHillers) [inline device_doctor/lib/src/device.dart:28]: Swarming bots are usually associated with one DUT. Is there a use case for Flutter to support multiple devices from one bot?

File: device_doctor/lib/src/device.dart [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+  /// Checks the health of the available devices.
+  Future<Map<String, HealthCheckResult>> checkDevices();
+
+  /// Recovers the device.
+  Future<void> recoverDevice();
```
Comment (CaseyHillers) [inline device_doctor/lib/src/device.dart:34]: The rest of the DeviceDiscovery API talks about devices, but this references one device. How do we know which device gets recovered?

File: device_doctor/lib/src/device.dart [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+/// A proxy for one specific device.
+abstract class Device {
```
Comment (CaseyHillers) [inline device_doctor/lib/src/device.dart:38]: This API seems too specific to implement beyond phones. Would `initialize``recover`, and `collect_logs` make more sense? Some of these methods make more sense as part of a child interface like `PhoneDevice`.

File: device_doctor/lib/src/device.dart [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  ///
+  /// Assumes the device doesn't have a secure unlock pattern.
+  Future<void> unlock();
+
+  /// Recovers the device back to a healthy state.
+  Future<void> recover();
+}
+
+/// Result of a health check for a specific parameter.
+class HealthCheckResult {
```
Comment (CaseyHillers) [inline device_doctor/lib/src/device.dart:70]: Health checks seem out of scope for this device file.

File: device_doctor/lib/src/process_helper.dart [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    _runningProcesses.remove(procInfo);
+  });
+
+  return proc;
+}
+
+/// Executes a command and returns its standard output as a String.
+///
+/// Standard error is redirected to the current process' standard error stream.
+Future<String> eval(String executable, List<String> arguments,
```
Comment (CaseyHillers) [inline device_doctor/lib/src/process_helper.dart:53]: I hesitate from `eval` as it's more associated with dynamic code calling, but here we know what we're calling. `execCommand`?

File: device_doctor/lib/src/utils.dart [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+  final int _level;
+  final String name;
+
+  static const LogLevel debug = const LogLevel._(0, 'DEBUG');
+  static const LogLevel info = const LogLevel._(1, 'INFO');
+  static const LogLevel warning = const LogLevel._(2, 'WARN');
+  static const LogLevel error = const LogLevel._(3, 'ERROR');
+}
+
+abstract class Logger {
```
Comment (CaseyHillers) [inline device_doctor/lib/src/utils.dart:28]: https://pub.dev/packages/logging

**Review by CaseyHillers — COMMENTED** [review]
Thank you for splitting this up!

We should add https://github.com/flutter/flutter/blob/master/analysis_options.yaml here to better match the Flutter style guide.

File: device_doctor/lib/src/config.dart [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  Config({
+    @required this.deviceOperatingSystem,
+  });
+
+  static void initialize(String deviceOS) {
```
Comment (jonahwilliams) [inline device_doctor/lib/src/config.dart:20]: Consider not using a global variable for passing around configuration. Its easy to track now when this project is small but will become difficult to refactor as it grows larger

File: device_doctor/lib/src/device.dart (hunk 8 above)
Comment (jonahwilliams) [inline device_doctor/lib/src/device.dart:28]: Might be more dart-y if this accepted a `Duration` object instead

File: device_doctor/lib/src/ios_device.dart [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+import 'package:meta/meta.dart';
+
+import 'device.dart';
+import 'process_helper.dart';
+
+/// IOS implementation of [DeviceDiscovery].
+///
+/// Discovers available ios devices and chooses one to work with.
+class IosDeviceDiscovery implements DeviceDiscovery {
+  factory IosDeviceDiscovery() {
```
Comment (jonahwilliams) [inline device_doctor/lib/src/ios_device.dart:17]: What is the goal of having this be a singleton?

File: device_doctor/lib/src/ios_device.dart [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  IosDeviceDiscovery._();
+
+  @visibleForTesting
+  IosDeviceDiscovery.testing();
+
+  static IosDeviceDiscovery _instance;
+
+  @override
+  Future<List<Device>> discoverDevices({int retriesDelayMs = 10000}) async {
+    final List<String> iosDeviceIds = LineSplitter.split(await deviceListOutput()).toList();
```
Comment (jonahwilliams) [inline device_doctor/lib/src/ios_device.dart:30]: you don't need to convert it to a List to map it, consider:

```
LineSplitter.split(await deviceListOutput())
  .map((String id) => IosDevice(deviceId: id))
  .toList()
```

File: device_doctor/lib/src/utils.dart [hunk 17]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:process/process.dart';
+
+/// Virtual current working directory, which affect functions, such as [exec].
+String cwd = Directory.current.path;
```
Comment (jonahwilliams) [inline device_doctor/lib/src/utils.dart:10]: If you use package path it has a `path.current` getter.

File: device_doctor/lib/src/utils.dart [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+ProcessManager processManager = LocalProcessManager();
```
Comment (jonahwilliams) [inline device_doctor/lib/src/utils.dart:12]: Same comment as elsewhere: globals are going to make this code difficult to refactor later

File: device_doctor/lib/src/utils.dart (hunk 13 above)
Comment (jonahwilliams) [reply device_doctor/lib/src/utils.dart:28]: Right, I would either use package logging, or a much simpler logging interface

File: device_doctor/bin/main.dart (hunk 1 above)
Comment (keyonghan) [reply device_doctor/bin/main.dart:9]: That seems for files located within lib?

File: device_doctor/bin/main.dart (hunk 2 above)
Comment (keyonghan) [reply device_doctor/bin/main.dart]: updated.

File: device_doctor/bin/main.dart (hunk 4 above)
Comment (keyonghan) [reply device_doctor/bin/main.dart]: Thanks for this info. Updated the parser with `help`.

File: device_doctor/bin/main.dart (hunk 3 above)
Comment (keyonghan) [reply device_doctor/bin/main.dart]: Added https://github.com/flutter/flutter/issues/66193

File: device_doctor/lib/src/config.dart (hunk 5 above)
Comment (keyonghan) [reply device_doctor/lib/src/config.dart]: Done.

File: device_doctor/lib/src/config.dart (hunk 6 above)
Comment (keyonghan) [reply device_doctor/lib/src/config.dart:13]: Yeah, I need to check swarming bot_config to see what we can use here, or maybe device id. Will include that once confirmed.

File: device_doctor/lib/src/config.dart (hunk 14 above)
Comment (keyonghan) [reply device_doctor/lib/src/config.dart]: Updated to use local variable instead.

File: device_doctor/lib/src/device.dart (hunk 7 above)
Comment (keyonghan) [reply device_doctor/lib/src/device.dart]: Good catch! Updated.

File: device_doctor/lib/src/device.dart (hunk 8 above)
Comment (keyonghan) [reply device_doctor/lib/src/device.dart]: We currently only support one device on one bot, but may support more to one in the future when needed.

Comment (keyonghan) [reply device_doctor/lib/src/device.dart]: Updated to `Duration`.

File: device_doctor/lib/src/device.dart (hunk 9 above)
Comment (keyonghan) [reply device_doctor/lib/src/device.dart]: Ah, a typo here. It should be `recoverDevices`. The implementation will recover all devices.
Will generalize this when `device id/ip` is available.

File: device_doctor/lib/src/device.dart (hunk 10 above)
Comment (keyonghan) [reply device_doctor/lib/src/device.dart:36]: The device here is meant for phone device. Updated the doc to reflect this.

File: device_doctor/lib/src/device.dart (hunk 11 above)
Comment (keyonghan) [reply device_doctor/lib/src/device.dart]: Moved to utils.

File: device_doctor/lib/src/ios_device.dart (hunk 15 above)
Comment (keyonghan) [reply device_doctor/lib/src/ios_device.dart:18]: This is mirrored over from existing cocoon/agent. My guess is to ensure only one `IosDeviceDiscovery` instance. @yjbanov may have a better idea?

File: device_doctor/lib/src/ios_device.dart (hunk 16 above)
Comment (keyonghan) [reply device_doctor/lib/src/ios_device.dart]: Thanks. Updated.

File: device_doctor/lib/src/process_helper.dart (hunk 12 above)
Comment (keyonghan) [reply device_doctor/lib/src/process_helper.dart:57]: As this is just the initial step of the tool, we will have more general cases on the way, and we also will implement exec as well.

File: device_doctor/lib/src/utils.dart (hunk 17 above)
Comment (keyonghan) [reply device_doctor/lib/src/utils.dart]: replaced.

File: device_doctor/lib/src/utils.dart (hunk 18 above)
Comment (keyonghan) [reply device_doctor/lib/src/utils.dart]: Changed to use local variables instead.

File: device_doctor/lib/src/utils.dart (hunk 13 above)
Comment (keyonghan) [reply device_doctor/lib/src/utils.dart]: Updated as suggested. Thanks!

File: device_doctor/lib/src/device.dart (hunk 8 above)
Comment (CaseyHillers) [reply device_doctor/lib/src/device.dart]: A host can run multiple swarming bots, but a swarming bot should only be connected to one DUT. Since this is a swarming bot utility, shouldn't this be phrased for a single device?

The counterexample I listed was moreso if there is a test that requires more than one phone. I've seen those set ups for swarming where they created specialized pools to support those use cases.

File: device_doctor/lib/src/device.dart (hunk 10 above)
Comment (CaseyHillers) [reply device_doctor/lib/src/device.dart:36]: My use case is how can we generalize this to also work with Fuchsia devices later on. With this approach, we'll need to further abstract this class when we do add that support.

File: device_doctor/lib/src/process_helper.dart (hunk 12 above)
Comment (CaseyHillers) [reply device_doctor/lib/src/process_helper.dart:57]: https://unix.stackexchange.com/questions/296838/whats-the-difference-between-eval-and-exec

What would be the difference between the exec and eval in this project?

Comment (keyonghan) [reply device_doctor/lib/src/process_helper.dart:57]: See https://github.com/flutter/cocoon/blob/master/agent/lib/src/utils.dart#L304 and https://github.com/flutter/cocoon/blob/master/agent/lib/src/utils.dart#L336

Comment (CaseyHillers) [reply device_doctor/lib/src/process_helper.dart:57]: Good point. Returning ProcessResult should deduplicate that logic

Comment (keyonghan) [reply device_doctor/lib/src/process_helper.dart:57]: Will consider consolidating when `exec` is called in ongoing PRs.

File: device_doctor/lib/src/device.dart (hunk 10 above)
Comment (keyonghan) [reply device_doctor/lib/src/device.dart:36]: Talked offline, will leave the class as is now. May update/extend the class when we get to the point to support fuchsia devices.

File: device_doctor/lib/src/config.dart [hunk 19]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
```
Comment (godofredoc) [inline device_doctor/lib/src/config.dart:5]: Do we actually need this class?  we can potentially directly create the classes in the factory method using the flags.

File: device_doctor/lib/src/device.dart [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+import 'config.dart';
+import 'ios_device.dart';
+import 'utils.dart';
+
+/// Operating system on the devices that this host is configured to test.
+enum DeviceOperatingSystem { android, ios }
+
+/// Discovers available devices and chooses one to work with.
+abstract class DeviceDiscovery {
+  factory DeviceDiscovery(Config config) {
```
Comment (godofredoc) [inline device_doctor/lib/src/device.dart:16]: We can use the flags directly in calls to this factory method.

File: device_doctor/lib/src/ios_device.dart [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+  @override
+  final String deviceId;
+
+  // The methods below are stubs for now. They will need to be expanded.
+  // We currently do not have a way to lock/unlock iOS devices. So we assume the
+  // devices are already unlocked. For now we'll just keep them at minimum
+  // screen brightness so they don't drain battery too fast.
+
+  @override
```
Comment (godofredoc) [inline device_doctor/lib/src/ios_device.dart:67]: The Device interface should not define all the methods below if they are not supported in ios only common methods.

File: device_doctor/lib/src/utils.dart [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+  BuildFailedError(this.message);
+
+  final String message;
+
+  @override
+  String toString() => message;
+}
+
+/// Result of a health check for a specific parameter.
+class HealthCheckResult {
```
Comment (godofredoc) [inline device_doctor/lib/src/utils.dart:23]: Does this class belong to a different file? health.dart maybe?

File: device_doctor/test/src/fake_ios_device.dart [hunk 23]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
```
Comment (godofredoc) [inline device_doctor/test/src/fake_ios_device.dart:1]: Why a fake rather than a mock?

File: device_doctor/bin/main.dart [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+    exit(1);
+  }
+
+  final Config config = Config(deviceOS: _deviceOS);
+  DeviceDiscovery(config);
+
+  // TODO(keyonghan): Implement healthcheck and recovery, https://github.com/flutter/flutter/issues/66193.
+}
+
+bool _checkArgs(ArgParser parser, List<String> args) {
```
[the remark this answers is not in the document]
Comment (keyonghan) [reply device_doctor/bin/main.dart:47]: Good point! Have implemented `callback` to handle this.

File: device_doctor/lib/src/config.dart (hunk 19 above)
Comment (keyonghan) [reply device_doctor/lib/src/config.dart:5]: Based on our current logic, this config class is not needed. Updated to use flag directly as suggested.

File: device_doctor/lib/src/device.dart (hunk 20 above)
Comment (keyonghan) [reply device_doctor/lib/src/device.dart:16]: Replaced.

File: device_doctor/lib/src/ios_device.dart (hunk 21 above)
Comment (keyonghan) [reply device_doctor/lib/src/ios_device.dart:67]: Removed the ones not common in ios.

File: device_doctor/lib/src/utils.dart (hunk 22 above)
Comment (keyonghan) [reply device_doctor/lib/src/utils.dart:23]: Moved to health.dart instead.

File: device_doctor/test/src/fake_ios_device.dart (hunk 23 above)
Comment (keyonghan) [reply device_doctor/test/src/fake_ios_device.dart:1]: This is used to test function `discoverDevices`, which calls another function `deviceListOutput`.
If we mock `discoverDevices` directly, we will lose testing logic of this function.
Instead if we use fake, `deviceListOut` can be manipulated so that we can test the logic of `discoverDevices`. This is how we do here.

**Review by godofredoc — APPROVED** [review]
LGTM but will defer the final approval to @CaseyHillers

File: device_doctor/lib/src/ios_device.dart [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+  Future<bool> isAwake() async => true;
+
+  @override
+  Future<bool> isAsleep() async => false;
+
+  @override
+  Future<void> unlock() async {}
```
Comment (CaseyHillers) [inline device_doctor/lib/src/ios_device.dart:74]: nit: In general, we should throw an unimplemented error on no-ops because we haven't finished the code

**Review by CaseyHillers — APPROVED** [review]
LGTM!

File: device_doctor/bin/main.dart [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+Future<void> main(List<String> args) async {
+  final ArgParser parser = ArgParser();
+  parser
+    ..addOption('$actionFlag', help: 'Supported actions are healthcheck and recovery.', callback: (String value) {
+      if (value == null || !supportedOptions.contains(value)) {
+        throw FormatException('Invalid value for option --action: $value');
+      }
+    })
+    ..addOption('$deviceOSFlag', help: 'Supported device OS: android and ios.', callback: (String value) {
+      if (value == null || !supportedDeviceOS.contains(value)) {
```
[the remark this answers is not in the document]
Comment (keyonghan) [reply device_doctor/bin/main.dart]: Good catch. The null check is not necessary. Removed.

File: device_doctor/lib/src/ios_device.dart (hunk 25 above)
Comment (keyonghan) [reply device_doctor/lib/src/ios_device.dart:74]: The following PR is intended to implement corresponding actions. Will leave it there.
